### PR TITLE
Update to python3 and django2

### PR DESCRIPTION
**Pull request by arthurian**
This PR adds support for python3 and django2. Also includes some other changes to improve documentation and test coverage. Specifically:

- Added unit tests.
- Updated documentation.
- Updated `setup.py` package metadata.
- Registered django admin models.
- Added info logging to trace the oauth flow.

@bermudezjd @elliottyates @sapnamysore @dodget Can you take a look when you get a chance?

### CODE REVIEW

File: README.rst
Comment (dodget) [inline README.rst]: Not to be nitpicky, but maybe this should be MIDDLEWARE

File: canvas_oauth/models.py
Comment (dodget) [inline canvas_oauth/models.py:50]: Possibly update string interpolation to use f-strings?

File: canvas_oauth/urls.py
Comment (dodget) [inline canvas_oauth/urls.py]: It is possible to just use `path` rather than `re_path` since there is no regular expression being evaluated? something like: `path('oauth-callback/', oauth_callback, name='canvas-oauth-callback')`

**Review by dodget — APPROVED** [review]
Looks good @arthurian!
I made a few middling comments, but nothing that impacts functionality.

File: README.rst
Comment (arthurian) [reply README.rst]: Yes, it should definitely be `MIDDLEWARE`. Good catch!

File: canvas_oauth/models.py
Comment (arthurian) [reply canvas_oauth/models.py:50]: Per our offline discussion, I think maybe we hold off updating to f-strings for now. I found two instances where I introduced f-strings, and I'll revert for now to stay consistent with the existing style. This will also retain python3.5 compatibility a little bit longer.

File: canvas_oauth/urls.py
Comment (arthurian) [reply canvas_oauth/urls.py]: Yes, definitely. The regex seems unnecessary here, so no reason not to use `path()`. I'll go ahead and update that.

**Comment by arthurian** [comment]
Bump. Any additional comments before merging from folks who haven't reviewed yet?

File: README.rst
Comment (elliottyates) [inline README.rst:21]: Ha! Maybe this should soon be... gasp... `zsh`!

File: README.rst
Comment (elliottyates) [inline README.rst:83]: This is a bit confusing to me, because the Canvas OAuth2 documentation doesn't show a param for this. Just not sure how it works without reading all the source; maybe I'm missing something. cf: https://canvas.instructure.com/doc/api/file.oauth.html#using-refresh-tokens

File: canvas_oauth/canvas.py
Comment (elliottyates) [inline canvas_oauth/canvas.py:32]: Just as a point of reference, you can also skip the `lambda` via https://docs.python.org/3/howto/sorting.html#operator-module-functions -- but this is probably a matter of personal preference (I certainly don't have a preference, this is just for fun!).

File: canvas_oauth/models.py
Comment (elliottyates) [reply canvas_oauth/models.py:50]: (although the docs say it requires python 3.6, fyi)

File: canvas_oauth/tests/test_models.py
Comment (elliottyates) [inline canvas_oauth/tests/test_models.py]: Nice that you at least have the `import` tested in `canvas_oauth/tests/test_oauth.py`!

File: canvas_oauth/urls.py
Comment (elliottyates) [inline canvas_oauth/urls.py:5]: Nice, I really hadn't used `path` yet...

File: requirements-dev.txt
Comment (elliottyates) [inline requirements-dev.txt]: Any reason these don't use the [compatible release operator](https://www.python.org/dev/peps/pep-0440/#version-specifiers) `~=`?

File: setup.py
Comment (elliottyates) [inline setup.py]: Are we sure we don't have a better group email, like `academictechnology@harvard.edu`?

File: tox.ini
Comment (elliottyates) [inline tox.ini:1]: Cool. :)

File: canvas_oauth/tests/test_oauth.py
Comment (elliottyates) [inline canvas_oauth/tests/test_oauth.py:242]: 😮 It might be nice for a future `mock`er to have a description of this magic with `type` and `PropertyMock`? It may be a common idiom, but I can't parse it at first glance. (It's used a few times in these tests.)

File: canvas_oauth/tests/test_settings.py
Comment (elliottyates) [inline canvas_oauth/tests/test_settings.py:17]: Wow. Cool. TBH I didn't really think you could just add and remove members of `settings` without somehow messing up Django. GTK. (In `pytest` you can do some of this using fixtures.)

**Review by elliottyates — APPROVED** [review]
These tests are amazing. Nice work @arthurian! It would awesome to see them run in github actions or travis, now that you have them, so that future PRs can be double-checked automatically. Thanks for showing me your changes, I learned a lot.

File: README.rst
Comment (arthurian) [reply README.rst:21]: Ha, maybe!

File: README.rst
Comment (arthurian) [reply README.rst:83]: I agree, this is a bit confusing on first glance, as it's not a standard parameter you will find in the Oauth2 docs. It's intended to force a refresh of the token before it actually expires (e.g. the `expires_in` provided in the access token response), providing a bit of a buffer so you don't acquire a token that is going to expire immediately. Perhaps it would be clearer to refactor this in terms of an absolute expiration instead of an offset/delta. For now, I'll try adding a bit more of an explanation.

File: canvas_oauth/canvas.py
Comment (arthurian) [reply canvas_oauth/canvas.py:32]: You're totally right, the `operator` module is a good alternative to the `lambda` here. I don't have a strong preference either way myself, but I'm glad to be reminded of that!

File: canvas_oauth/tests/test_models.py
Comment (arthurian) [reply canvas_oauth/tests/test_models.py]: Good catch. Need to add some tests here.

File: requirements-dev.txt
Comment (arthurian) [reply requirements-dev.txt]: The compatible release operator would be great here. No reason not to use it, so I'm going to fix that, thanks!

File: setup.py
Comment (arthurian) [reply setup.py]: Good question, I'm not aware if we have a better group email, but I will check with some folks to see if there is.

File: canvas_oauth/tests/test_oauth.py
Comment (arthurian) [reply canvas_oauth/tests/test_oauth.py:242]: Good point, it's a bit of an odd idiom that's new to me, and maybe too obscure to be worth using. In any event, I'll add a comment to that affect for future code readers, taken from the unittest documentation: 
 https://docs.python.org/3/library/unittest.mock.html#unittest.mock.PropertyMock

File: canvas_oauth/tests/test_settings.py
Comment (arthurian) [reply canvas_oauth/tests/test_settings.py:17]: GTK that this might be more easily accomplished in `pytest` with some basic fixtures.

File: canvas_oauth/tests/test_oauth.py
Comment (elliottyates) [reply canvas_oauth/tests/test_oauth.py:242]: Yeah I think in part it's the `type()` that's making my head hurt here. Good to see it's directly in the `unittest` docs.

File: setup.py
Comment (arthurian) [reply setup.py]: Yes, you're right, looks like that's what we should be using based on some of our other projects. Thanks for the tip!

**Comment by arthurian** [comment]
Thanks all for the comments and suggestions! I'm moving ahead with the merge.

**Pull request merged**
